File: src/query/classes/RedisCache.ts
// src/query/classes/RedisCache.ts

import { createClient, RedisClientType } from 'redis';
import { JSONStringify, logger } from '../../utils/utils';

class RedisCacheClass {
    private client: RedisClientType | null = null;
    private keyPrefix: string;
    private debugLogs: boolean;
    private redisUrl: string | null | undefined = null;

    constructor(keyPrefix: string, debugLogs: boolean = false) {
        this.keyPrefix = keyPrefix;
        this.debugLogs = debugLogs;
        this.initializeClient();
    }

    private initializeClient() {
        this.redisUrl = process.env.JSINFO_QUERY_REDDIS_CACHE;
        if (!this.redisUrl) {
            this.log('JSINFO_QUERY_REDDIS_CACHE environment variable is not set.');
            return;
        }
        this.connect();
    }

    private async connect() {
        if (!this.redisUrl) return;
        this.log('Attempting to reconnect to Redis...');
        this.client = createClient({
            url: this.redisUrl,
            socket: {
                connectTimeout: 5000, // Timeout for connecting to Redis in milliseconds
            },
        });
        this.client.on('error', (err) => this.logError('Redis Client Error', err));
        this.client.connect().catch((err) => {
            this.logError('Failed to connect to Redis', err);
            this.client = null;
        });
    }

    private async reconnect() {
        try {
            this.connect();
        } catch (error) {
            this.logError(`Reddis reconnect failed`, error);
        }
    }

    async get(key: string): Promise<string | null> {
        if (!this.client) {
            this.log('Redis client is not available.');
            return null;
        }

        const fullKey = this.keyPrefix + key;
        const startTime = Date.now();
        try {
            const promise = this.client.get(fullKey);
            const timeout = new Promise((_, reject) => setTimeout(() => reject(new Error('Timeout')), 2000));
            const result = await Promise.race([promise, timeout]);
            const endTime = Date.now();
            const timeTaken = endTime - startTime;

            if (!result) {
                this.log(`Cache miss for key ${fullKey}. Time taken: ${timeTaken}ms`);
                return null;
            } else {
                this.log(`Cache hit for key ${fullKey}. Time taken: ${timeTaken}ms`);
                return String(result);
            }
        } catch (error) {
            const endTime = Date.now();
            const timeTaken = endTime - startTime;
            this.logError(`Error getting key ${fullKey} from Redis. Time taken: ${timeTaken}ms`, error);
            await this.reconnect();
            return null;
        }
    }

    async set(key: string, value: string, ttl: number = 30): Promise<void> {
        if (!this.client) {
            this.log('Redis client is not available.');
            return;
        }

        try {
            await this.client.set(this.keyPrefix + key, value, {
                EX: ttl,
            });
        } catch (error) {
            this.logError(`Error setting key ${this.keyPrefix + key} in Redis`, error);
            await this.reconnect();
        }
    }

    async getArray(key: string): Promise<any[] | null> {
        const result = await this.get(key);
        if (!result) return null;
        try {
            return JSON.parse(result);
        } catch (error) {
            this.logError(`Error parsing JSON for key ${this.keyPrefix + key}`, error);
            return null;
        }
    }

    async setArray(key: string, value: any[], ttl: number = 30): Promise<void> {
        try {
            const stringValue = JSONStringify(value);
            await this.set(key, stringValue, ttl);
        } catch (error) {
            this.logError(`Error serializing array for key ${this.keyPrefix + key}`, error);
        }
    }

    async getDict(key: string): Promise<{ [key: string]: any } | null> {
        const result = await this.get(key);
        if (!result) return null;
        try {
            return JSON.parse(result);
        } catch (error) {
            this.logError(`Error parsing JSON for key ${this.keyPrefix + key}`, error);
            return null;
        }
    }

    async setDict(key: string, value: { [key: string]: any }, ttl: number = 30): Promise<void> {
        try {
            const stringValue = JSON.stringify(value);
            await this.set(key, stringValue, ttl);
        } catch (error) {
            this.logError(`Error serializing dictionary for key ${this.keyPrefix + key}`, error);
        }
    }
    private log(message: string) {
        if (!this.debugLogs) return;
        logger.info(`RedisCache: ${message}`);
    }

    private logError(message: string, error: any) {
        logger.error(`RedisCache: ${message}`, error);
    }
}

export const RedisCache = new RedisCacheClass("jsinfo");

File: lavapProviderHealth/rediscache.py
import redis
import logging
import env
from utils import convert_dict_to_dbjson, json_load_or_none

class RedisCacheClass:
    def __init__(self, key_prefix: str, debug_logs: bool = False):
        self.key_prefix = key_prefix
        self.debug_logs = debug_logs
        self.redis_url = None
        self.client = None
        self.initialize_client()

    def initialize_client(self):
        self.redis_url = env.REDIS_SERVER
        if not self.redis_url:
            self.log('JSINFO_QUERY_REDDIS_CACHE environment variable is not set.')
            return
        self.connect()

    def connect(self):
        if not self.redis_url:
            return
        self.log('Attempting to reconnect to Redis...')
        try:
            self.client = redis.Redis.from_url(self.redis_url, socket_connect_timeout=5)  # 5 seconds timeout
            self.client.ping()  # Check connection
        except redis.RedisError as err:
            self.log_error('Failed to connect to Redis', err)
            self.client = None

    def reconnect(self):
        try:
            self.connect()
        except Exception as error:
            self.log_error('Redis reconnect failed', error)

    def get(self, key: str):
        if not self.client:
            self.log('Redis client is not available.')
            return None

        full_key = self.key_prefix + key
        try:
            result = self.client.get(full_key)
            if result is None:
                self.log(f'Cache miss for key {full_key}.')
                return None
            else:
                self.log(f'Cache hit for key {full_key}.')
                return result.decode('utf-8')
        except Exception as error:
            self.log_error(f'Error getting key {full_key} from Redis.', error)
            self.reconnect()
            return None

    def set(self, key: str, value: str, ttl: int = 30):
        if not self.client:
            self.log('Redis client is not available.')
            return

        try:
            self.client.setex(self.key_prefix + key, ttl, value)
        except Exception as error:
            self.log_error(f'Error setting key {self.key_prefix + key} in Redis', error)
            self.reconnect()

    def get_array(self, key: str) -> list | None:
        result = self.get(key)
        if not result:
            return None
        return json_load_or_none(result)

    def set_array(self, key: str, value: list, ttl: int = 30):
        try:
            string_value = convert_dict_to_dbjson(value)
            self.set(key, string_value, ttl)
        except Exception as error:
            self.log_error(f'Error serializing array for key {self.key_prefix + key}', error)

    def get_dict(self, key: str) -> dict | None:
        result = self.get(key)
        if not result:
            return None
        return json_load_or_none(result, )

    def set_dict(self, key: str, value: dict, ttl: int = 30):
        try:
            string_value = convert_dict_to_dbjson(value)
            self.set(key, string_value, ttl)
        except Exception as error:
            self.log_error(f'Error serializing dict for key {self.key_prefix + key}', error)
            
    def lpush_dict(self, key: str, value: dict):
        if not self.client:
            self.log('Redis client is not available.')
            return

        full_key = self.key_prefix + key
        try:            
            self.client.lpush(full_key, convert_dict_to_dbjson(value))
        except Exception as error:
            self.log_error(f'Error LPUSH to key {full_key} in Redis', error)
            self.reconnect()

    def brpop_dict(self, key: str, timeout: int = 0) -> dict | None:
        if not self.client:
            self.log('Redis client is not available.')
            return None

        full_key = self.key_prefix + key
        try:
            result = self.client.brpop(full_key, timeout)
            if result is None:
                self.log(f'BRPOP timeout or no elements for key {full_key}.')
                return None
            else:
                # result is a tuple (key, value), where value is the element popped
                return json_load_or_none(result[1])
        except Exception as error:
            self.log_error(f'Error BRPOP from key {full_key} in Redis.', error)
            self.reconnect()
            return None
    
    def log(self, message: str):
        if not self.debug_logs:
            return
        logging.info(f'RedisCache: {message}')

    def log_error(self, message: str, error):
        logging.error(f'RedisCache: {message}', exc_info=error)

rediscache = RedisCacheClass("jsinfo-healthp")